phylum/connector.lisp

```diff
@@ -0,0 +1,215 @@
+;; for processing the request. The context also stores the OID, which is used
+;; when processing the response.
+;;
+;; The request body itself is stored either in the stateDB or in a PDC, and is
+;; referenced indirectly by the event context. The body contains the request
+;; payload that is to be forwarded to the third-party system. Every request has
+;; a unique request ID, which is used to correlate a response to a callback.
+;; *IMPORTANT*: this request ID is NOT the same as the request ID embedded
+;; in the transaction context, which is primarily used for tracing.
+;;
+;; *Response Handling Logic*:
+;;
+;; Once the hub has forwarded the request, and received a response from the
+;; third-party system, it then sends this response to the `$ch_callback `
+;; entrypoint, passing in the response as transient data using key prefix 
+;; `$ch_rep:N`, which may include multiple responses for different requests.
+;; This response includes the request ID, which is used to correlate the
+;; response with the request via the same event context. The MSPID in the
+;; original event context is used to ensure that a response is only received
+;; by the correct org, so an org cannot spoof responses for requests from
+;; another org.
+;;
+;; The phylum logic for `$ch_callback` uses the req ID to lookup the event
+;; context, along with the OID and class factory. The class factory is used
+;; to instiante the object for that OID, and call the handle method on that
+;; object with the response.
+;;
+;; Upon execution of the callback, the original event context is purged, along
+;; with the original request (or deleted if in state DB), so as to reduce
+;; storage space.
+;;
+(defun mk-connector-handler ()
+  (let ([state (sorted-map)])
+    (labels
+      ([register-handler (name fn)
+         (assoc! state name fn)]
+
+       [mk-request-key (req-id)
+         (format-string "$cr:{}" req-id)]
+
+       ;; IMPORTANT: we store this in the central `private` PDC since we
+       ;; don't know which PDC the original request was stored.
+       [register-request-callback (req-id handler-name &optional ctx)
+         (let* ([ctx (default ctx (sorted-map))]
+                [callback-state (sorted-map
+                               "handler_name" handler-name)])
+           (assoc! ctx "request_id" req-id)
+           (assoc! callback-state "ctx" ctx)
+           (sidedb:put (mk-request-key req-id) callback-state))]
+
+       [unregister-request (req-id ctx)
+         ;; TODO: handle multiple responses for single request_id (optional) 
+         (let* ([callback-state-key (mk-request-key req-id)]
+                [key (get ctx "key")]
+                [pdc (get ctx "pdc")])
+           (sidedb:purge callback-state-key)
+           (if pdc
+            (cc:storage-purge-private pdc key)
+            (statedb:del key)))]
+
+       [get-callback-state (req-id)
+         (sidedb:get (or (mk-request-key req-id)
```

Comment:
Should this `or` be around the `get`?

Reply:
good catch ,will fix

Reply:
I think the parens look a bit unbalanced after this change.